File: backend/internal/service/dosen_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"time"

	"topsus-krs/internal/domain"
	"topsus-krs/internal/dto"
	"topsus-krs/internal/repository"
	"topsus-krs/pkg/cache"
	"topsus-krs/pkg/pagination"
)

// DosenService mendefinisikan kontrak business logic untuk dosen
type DosenService interface {
	GetAll(ctx context.Context, params *pagination.Params) ([]*dto.DosenResponse, int64, error)
	GetByID(ctx context.Context, id uint64) (*dto.DosenResponse, error)
	GetMahasiswaBimbingan(ctx context.Context, dosenID uint64, params *pagination.Params) ([]*dto.MahasiswaResponse, int64, error)
	Create(ctx context.Context, req *dto.CreateDosenRequest) (*dto.DosenResponse, error)
	Update(ctx context.Context, id uint64, req *dto.UpdateDosenRequest) (*dto.DosenResponse, error)
	Delete(ctx context.Context, id uint64) error
}

type dosenService struct {
	repo          repository.DosenRepository
	mahasiswaRepo repository.MahasiswaRepository
	cache         *cache.RedisCache
}

func NewDosenService(
	repo repository.DosenRepository,
	mahasiswaRepo repository.MahasiswaRepository,
	redisCache *cache.RedisCache,
) DosenService {
	return &dosenService{
		repo:          repo,
		mahasiswaRepo: mahasiswaRepo,
		cache:         redisCache,
	}
}

func (s *dosenService) GetAll(ctx context.Context, params *pagination.Params) ([]*dto.DosenResponse, int64, error) {
	cacheKey := fmt.Sprintf("dosen:list:p=%d:l=%d:q=%s", params.Page, params.Limit, params.Search)

	type cachedList struct {
		Items []*dto.DosenResponse `json:"items"`
		Total int64                `json:"total"`
	}

	var cached cachedList
	if err := s.cache.Get(ctx, cacheKey, &cached); err == nil {
		return cached.Items, cached.Total, nil
	}

	dosens, total, err := s.repo.FindAll(ctx, params)
	if err != nil {
		return nil, 0, err
	}

	items := make([]*dto.DosenResponse, 0, len(dosens))
	for _, d := range dosens {
		items = append(items, mapDosenToDTO(d))
	}

	_ = s.cache.Set(ctx, cacheKey, cachedList{Items: items, Total: total}, 5*time.Minute)

	return items, total, nil
}

func (s *dosenService) GetByID(ctx context.Context, id uint64) (*dto.DosenResponse, error) {
	cacheKey := fmt.Sprintf("dosen:%d", id)

	var cached dto.DosenResponse
	if err := s.cache.Get(ctx, cacheKey, &cached); err == nil {
		return &cached, nil
	}

	d, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if d == nil {
		return nil, errors.New("dosen tidak ditemukan")
	}

	result := mapDosenToDTO(d)
	_ = s.cache.Set(ctx, cacheKey, result, 10*time.Minute)

	return result, nil
}

func (s *dosenService) GetMahasiswaBimbingan(ctx context.Context, dosenID uint64, params *pagination.Params) ([]*dto.MahasiswaResponse, int64, error) {
	// Pastikan dosen ada
	dosen, err := s.repo.FindByID(ctx, dosenID)
	if err != nil {
		return nil, 0, err
	}
	if dosen == nil {
		return nil, 0, errors.New("dosen tidak ditemukan")
	}

	cacheKey := fmt.Sprintf("dosen:%d:mahasiswa:p=%d:l=%d:q=%s", dosenID, params.Page, params.Limit, params.Search)

	type cachedList struct {
		Items []*dto.MahasiswaResponse `json:"items"`
		Total int64                    `json:"total"`
	}

	var cached cachedList
	if err := s.cache.Get(ctx, cacheKey, &cached); err == nil {
		return cached.Items, cached.Total, nil
	}

	mahasiswas, total, err := s.mahasiswaRepo.FindByDosenPA(ctx, dosenID, params)
	if err != nil {
		return nil, 0, err
	}

	items := make([]*dto.MahasiswaResponse, 0, len(mahasiswas))
	for _, m := range mahasiswas {
		items = append(items, mapMahasiswaToDTO(m))
	}

	_ = s.cache.Set(ctx, cacheKey, cachedList{Items: items, Total: total}, 5*time.Minute)

	return items, total, nil
}

func (s *dosenService) Create(ctx context.Context, req *dto.CreateDosenRequest) (*dto.DosenResponse, error) {
	existing, err := s.repo.FindByNIDN(ctx, req.NIDN)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return nil, errors.New("NIDN sudah terdaftar")
	}

	existingEmail, err := s.repo.FindByEmail(ctx, req.Email)
	if err != nil {
		return nil, err
	}
	if existingEmail != nil {
		return nil, errors.New("email dosen sudah terdaftar")
	}

	d := &domain.Dosen{
		UserID: req.UserID,
		NIDN:   req.NIDN,
		Nama:   req.Nama,
		Email:  req.Email,
		Prodi:  req.Prodi,
	}

	if err := s.repo.Create(ctx, d); err != nil {
		return nil, err
	}

	s.invalidateListCache(ctx)

	return mapDosenToDTO(d), nil
}

func (s *dosenService) Update(ctx context.Context, id uint64, req *dto.UpdateDosenRequest) (*dto.DosenResponse, error) {
	d, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if d == nil {
		return nil, errors.New("dosen tidak ditemukan")
	}

	if req.NIDN != "" && req.NIDN != d.NIDN {
		existing, err := s.repo.FindByNIDN(ctx, req.NIDN)
		if err != nil {
			return nil, err
		}
		if existing != nil && existing.ID != id {
			return nil, errors.New("NIDN sudah digunakan dosen lain")
		}
		d.NIDN = req.NIDN
	}

	if req.Email != "" && req.Email != d.Email {
		existing, err := s.repo.FindByEmail(ctx, req.Email)
		if err != nil {
			return nil, err
		}
		if existing != nil && existing.ID != id {
			return nil, errors.New("email sudah digunakan dosen lain")
		}
		d.Email = req.Email
	}

	if req.Nama != "" {
		d.Nama = req.Nama
	}
	if req.Prodi != "" {
		d.Prodi = req.Prodi
	}

	if err := s.repo.Update(ctx, d); err != nil {
		return nil, err
	}

	s.invalidateListCache(ctx)
	_ = s.cache.Delete(ctx, fmt.Sprintf("dosen:%d", id))

	return mapDosenToDTO(d), nil
}

func (s *dosenService) Delete(ctx context.Context, id uint64) error {
	d, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return err
	}
	if d == nil {
		return errors.New("dosen tidak ditemukan")
	}

	if err := s.repo.Delete(ctx, id); err != nil {
		return err
	}

	s.invalidateListCache(ctx)
	_ = s.cache.Delete(ctx, fmt.Sprintf("dosen:%d", id))

	return nil
}

func (s *dosenService) invalidateListCache(ctx context.Context) {
	_ = s.cache.DeleteByPattern(ctx, "dosen:list:*")
	_ = s.cache.Delete(ctx, "dashboard:stats")
}

func mapDosenToDTO(d *domain.Dosen) *dto.DosenResponse {
	return &dto.DosenResponse{
		ID:        d.ID,
		UserID:    d.UserID,
		NIDN:      d.NIDN,
		Nama:      d.Nama,
		Email:     d.Email,
		Prodi:     d.Prodi,
		CreatedAt: d.CreatedAt,
		UpdatedAt: d.UpdatedAt,
	}
}

File: backend-refactor-python/app/services/dosen_service.py
from fastapi import HTTPException
from sqlalchemy.orm import Session

from ..redis_client import cache
from ..repositories.dosen_repository import DosenRepository
from ..repositories.mahasiswa_repository import MahasiswaRepository
from ..schemas.dosen import CreateDosenRequest, DosenResponse, UpdateDosenRequest
from ..schemas.mahasiswa import MahasiswaResponse

_TTL_LIST = 5 * 60
_TTL_ITEM = 10 * 60


class DosenService:
    def __init__(self, db: Session):
        self.repo          = DosenRepository(db)
        self.mhs_repo      = MahasiswaRepository(db)

    def get_all(self, page: int, limit: int, search: str) -> tuple[list[DosenResponse], int]:
        key = f"dosen:list:p={page}:l={limit}:q={search}"
        cached = cache.get(key)
        if cached:
            items = [DosenResponse(**i) for i in cached["items"]]
            return items, cached["total"]

        rows, total = self.repo.find_all(page, limit, search)
        items = [DosenResponse(**r) for r in rows]
        cache.set(key, {"items": [i.model_dump(mode="json") for i in items], "total": total}, _TTL_LIST)
        return items, total

    def get_by_id(self, id: int) -> DosenResponse:
        key = f"dosen:{id}"
        cached = cache.get(key)
        if cached:
            return DosenResponse(**cached)

        row = self.repo.find_by_id(id)
        if not row:
            raise HTTPException(status_code=404, detail="Dosen tidak ditemukan")
        result = DosenResponse(**row)
        cache.set(key, result.model_dump(mode="json"), _TTL_ITEM)
        return result

    def get_mahasiswa_bimbingan(
        self, dosen_id: int, page: int, limit: int, search: str
    ) -> tuple[list[MahasiswaResponse], int]:
        if not self.repo.find_by_id(dosen_id):
            raise HTTPException(status_code=404, detail="Dosen tidak ditemukan")

        key = f"dosen:{dosen_id}:mahasiswa:p={page}:l={limit}:q={search}"
        cached = cache.get(key)
        if cached:
            items = [MahasiswaResponse(**i) for i in cached["items"]]
            return items, cached["total"]

        rows, total = self.mhs_repo.find_by_dosen_pa(dosen_id, page, limit, search)
        items = [MahasiswaResponse(**r) for r in rows]
        cache.set(key, {"items": [i.model_dump(mode="json") for i in items], "total": total}, _TTL_LIST)
        return items, total

    def create(self, req: CreateDosenRequest) -> DosenResponse:
        if self.repo.find_by_nidn(req.nidn):
            raise HTTPException(status_code=409, detail="NIDN sudah terdaftar")
        if self.repo.find_by_email(req.email):
            raise HTTPException(status_code=409, detail="Email dosen sudah terdaftar")

        data = {
            "user_id": req.user_id,
            "nidn":    req.nidn,
            "nama":    req.nama,
            "email":   req.email,
            "prodi":   req.prodi,
        }
        new_id = self.repo.create(data)
        self._invalidate_cache()
        row = self.repo.find_by_id(new_id)
        return DosenResponse(**row)

    def update(self, id: int, req: UpdateDosenRequest) -> DosenResponse:
        row = self.repo.find_by_id(id)
        if not row:
            raise HTTPException(status_code=404, detail="Dosen tidak ditemukan")

        nidn  = req.nidn  if req.nidn  else row["nidn"]
        nama  = req.nama  if req.nama  else row["nama"]
        email = req.email if req.email else row["email"]
        prodi = req.prodi if req.prodi else row["prodi"]

        if nidn != row["nidn"] and self.repo.find_by_nidn(nidn):
            raise HTTPException(status_code=409, detail="NIDN sudah digunakan dosen lain")
        if email != row["email"] and self.repo.find_by_email(email):
            raise HTTPException(status_code=409, detail="Email sudah digunakan dosen lain")

        self.repo.update(id, {"nidn": nidn, "nama": nama, "email": email, "prodi": prodi})
        self._invalidate_cache()
        cache.delete(f"dosen:{id}")
        row = self.repo.find_by_id(id)
        return DosenResponse(**row)

    def delete(self, id: int) -> None:
        if not self.repo.find_by_id(id):
            raise HTTPException(status_code=404, detail="Dosen tidak ditemukan")
        self.repo.delete(id)
        self._invalidate_cache()
        cache.delete(f"dosen:{id}")

    def _invalidate_cache(self):
        cache.delete_pattern("dosen:list:*")
        cache.delete("dashboard:stats")
